manpages/csetup.1
.TH CSETUP 1 "2026-04-08" "" ""
.SH NAME
csetup \- Workspace setup tool for cbuild
.SH SYNOPSIS
.B csetup
<\fIsubcommand\fR>
.SH DESCRIPTION
Workspace setup tool for cbuild
.SH GLOBAL OPTIONS
.TP
.B \-w <value>, \-\-workspace <value>
Path to the workspace directory.
.TP
.B \-h, \-\-help
Show help for the current command. Prefer the `help` subcommand for command-specific help.
.SH SUBCOMMANDS
.SS add-config
Add a build configuration

.B Synopsis:
csetup add-config

.B Options for add-config:
.TP
.B \-c <value>, \-\-config <value>
Build configuration to use (e.g., Debug, Release), comma separated
.TP
.B \-T <value>, \-\-toolchain <value>
Select toolchain to use.
.SS add-dependency
Add a dependency to a target

.B Synopsis:
csetup add-dependency <target> <dependency>

.SS add-testing-dependency
Add a testing dependency to a target

.B Synopsis:
csetup add-testing-dependency <target> <dependency>

.SS declare-git-source
Add git source information to the workspace without downloading

.B Synopsis:
csetup declare-git-source <url> [source]

.SS declare-local-source
Add local source information to the workspace

.B Synopsis:
csetup declare-local-source <source> <sourcepath>

.B Options for declare-local-source:
.TP
.B \-\-sourcepath <value>
path to the local source
.SS detect-toolchains
Detect system toolchains

.B Synopsis:
csetup detect-toolchains

.B Options for detect-toolchains:
.TP
.B \-\-debug
Show debug information.
.SS dev-init
Initialize a new workspace for development in the current directory

.B Synopsis:
csetup dev-init [workspace]

.B Options for dev-init:
.TP
.B \-\-download-deps
Download dependencies during clone.
.TP
.B \-\-no-setup
Don't run setup after downloading or cloning.
.SS disable-staging
Disable staging for a target

.B Synopsis:
csetup disable-staging <target> <target>

.B Options for disable-staging:
.TP
.B \-t <value>, \-\-target <value>
Specific target to build/modify.
.SS disable-testing
Disable testing for a target

.B Synopsis:
csetup disable-testing <target>

.SS download
Download missing sources

.B Synopsis:
csetup download [source]

.B Options for download:
.TP
.B \-\-download-deps
Download dependencies during clone.
.TP
.B \-\-no-setup
Don't run setup after downloading or cloning.
.TP
.B \-\-submodule
Add as a git submodule instead of cloning.
.SS drop-files
Delete local source files without removing them from configuration

.B Synopsis:
csetup drop-files [source]

.SS enable-staging
Enable staging for a target

.B Synopsis:
csetup enable-staging <target> <target>

.B Options for enable-staging:
.TP
.B \-t <value>, \-\-target <value>
Specific target to build/modify.
.SS enable-testing
Enable testing for a target

.B Synopsis:
csetup enable-testing <target>

.SS get-args
Get build arguments for a target

.B Synopsis:
csetup get-args <target> <target>

.B Options for get-args:
.TP
.B \-c <value>, \-\-config <value>
Build configuration to use (e.g., Debug, Release), comma separated
.TP
.B \-T <value>, \-\-toolchain <value>
Select toolchain to use.
.TP
.B \-t <value>, \-\-target <value>
Specific target to build/modify.
.SS git-clone
Clone a git repository into the workspace

.B Synopsis:
csetup git-clone <url> [source]

.B Options for git-clone:
.TP
.B \-\-download-deps
Download dependencies during clone.
.TP
.B \-\-submodule
Add as a git submodule instead of cloning.
.TP
.B \-\-no-setup
Don't run setup after downloading or cloning.
.TP
.B \-\-branch <value>
Branch name to use.
.TP
.B \-\-revision <value>
Revision or tag to use.
.TP
.B \-\-no-track
Do not track the cloned branch.
.SS help
Show help for csetup or a specific subcommand

.B Synopsis:
csetup help [subcommand]

.SS init
Initialize a new workspace

.B Synopsis:
csetup init <workspace>

.B Options for init:
.TP
.B \-\-reinit
Reinitialize the workspace.
.SS list-sources
List all sources in the workspace

.B Synopsis:
csetup list-sources

.SS load-defaults
Load default configuration for a source from its csetup.yml

.B Synopsis:
csetup load-defaults <source>

.SS new-target
Add a new target to the workspace

.B Synopsis:
csetup new-target [target] <source>

.B Options for new-target:
.TP
.B \-s <value>, \-\-source <value>
Specific source to modify.
.TP
.B \-t <value>, \-\-target <value>
Specific target to build/modify.
.TP
.B \-\-overwrite
Overwrite target if it exists.
.TP
.B \-p <value>, \-\-project-type <value>
The project type (e.g., CMake).
.TP
.B \-\-cmake-package-name <value>
The CMake package name.
.SS pin
Pin a git source (HEAD by default, or a specific revision)

.B Synopsis:
csetup pin [source]

.B Options for pin:
.TP
.B \-\-revision <value>
Revision or tag to use.
.SS remove-config
Remove a build configuration

.B Synopsis:
csetup remove-config

.B Options for remove-config:
.TP
.B \-c <value>, \-\-config <value>
Build configuration to use (e.g., Debug, Release), comma separated
.SS remove-dependency
Remove a dependency from a target

.B Synopsis:
csetup remove-dependency <target> <dependency>

.SS remove-project
Remove a source and all its associated targets from the workspace

.B Synopsis:
csetup remove-project <source>

.B Options for remove-project:
.TP
.B \-X, \-\-delete
Delete files when removing source.
.TP
.B \-s <value>, \-\-source <value>
Specific source to modify.
.SS remove-source
Remove a source from the workspace

.B Synopsis:
csetup remove-source <source>

.B Options for remove-source:
.TP
.B \-X, \-\-delete
Delete files when removing source.
.TP
.B \-s <value>, \-\-source <value>
Specific source to modify.
.SS remove-target
Remove a target from the workspace

.B Synopsis:
csetup remove-target <target>

.B Options for remove-target:
.TP
.B \-t <value>, \-\-target <value>
Specific target to build/modify.
.SS remove-testing-dependency
Remove a testing dependency from a target

.B Synopsis:
csetup remove-testing-dependency <target> <dependency>

.SS set-cxx-version
Set the C++ version for a target or the whole workspace

.B Synopsis:
csetup set-cxx-version <version> [target]

.B Options for set-cxx-version:
.TP
.B \-t <value>, \-\-target <value>
Specific target to build/modify.
.SS status
Show status for sources (clean/modified, HEAD, expected revision)

.B Synopsis:
csetup status [source]

.SS tidy
Delete source folders that are not in the sources list

.B Synopsis:
csetup tidy

.B Options for tidy:
.TP
.B \-d, \-\-dry-run
Show commands without executing them.
.SS track
Track a branch for a git source

.B Synopsis:
csetup track <source> <branch>

.SS unpin
Unpin a git source (clear revision)

.B Synopsis:
csetup unpin <source>

.SS untrack
Stop tracking a branch for a git source

.B Synopsis:
csetup untrack <source>

.SS update
Update a git source (pull latest or move to revision)

.B Synopsis:
csetup update <source>

.B Options for update:
.TP
.B \-\-revision <value>
Revision or tag to use.
.SS versions
List pinned/unpinned versions for sources

.B Synopsis:
csetup versions [source]



app/csetupapp/csetup.go
package csetupapp

import (
	"context"
	"fmt"
	"os"
	"path/filepath"

	"gitlab.com/rpnx/cbuild-go/pkg/ccommon"
	"gitlab.com/rpnx/cbuild-go/pkg/cli"
)

func getWorkspacePath(ctx context.Context) string {
	workspacePathRaw := cli.GetOptionalPath(ctx, ccommon.WorkspaceParameter)
	workspacePath := ""
	if workspacePathRaw != nil {
		workspacePath = *workspacePathRaw
	}
	if workspacePath == "" {
		workspacePath = "."
	}

	// Try to find the workspace by looking up from current directory
	if workspacePath == "." {
		cwd, err := os.Getwd()
		if err == nil {
			curr := cwd
			for {
				if _, err := os.Stat(filepath.Join(curr, "cbuild_workspace.yml")); err == nil {
					absPath, err := filepath.Abs(curr)
					if err != nil {
						fmt.Fprintf(os.Stderr, "Error: failed to get absolute path: %v\n", err)
						os.Exit(1)
					}
					workspacePath = absPath
					break
				}
				parent := filepath.Dir(curr)
				if parent == curr {
					break
				}
				curr = parent
			}
		}
	} else {
		absPath, err := filepath.Abs(workspacePath)
		if err == nil {
			workspacePath = absPath
		}
	}
	return workspacePath
}

var CSetup = &cli.Runner{
	Name:        "csetup",
	Description: "Workspace setup tool for cbuild",
	GlobalFlags: []cli.Flag{
		ccommon.WorkspaceFlag,
		ccommon.HelpFlag,
	},
	Subcommands: make(map[string]*cli.Subcommand),
}

var (
	URLParameter            = cli.Parameter{Key: "url", Type: cli.ParameterTypeURI, DefaultValue: nil, Description: "URL of the repository"}
	SourcePathParameter     = cli.Parameter{Key: "sourcepath", Type: cli.ParameterTypePath, DefaultValue: nil, Description: "path to the local source"}
	SourcePathFlag          = cli.Flag{Short: "", Long: "sourcepath", Parameter: SourcePathParameter}
	HelpSubcommandParameter = cli.Parameter{Key: "help-subcommand", Type: cli.ParameterTypeString, DefaultValue: nil, Description: "subcommand to show help for"}
	SourceParameter         = cli.Parameter{Key: "source", Type: cli.ParameterTypeString, DefaultValue: nil, Description: "name of the source"}
	DependencyParameter     = cli.Parameter{Key: "dependency", Type: cli.ParameterTypeString, DefaultValue: nil, Description: "name of the dependency"}
	TargetParameter         = cli.Parameter{Key: "target", Type: cli.ParameterTypeString, DefaultValue: nil, Description: "Name of the build target."}
	RevisionParameter       = cli.Parameter{Key: "revision", Type: cli.ParameterTypeString, DefaultValue: nil, Description: "Revision or tag to use."}
	BranchParameter         = cli.Parameter{Key: "branch", Type: cli.ParameterTypeString, DefaultValue: nil, Description: "Branch name to use."}
	NoTrackParameter        = cli.Parameter{Key: "no-track", Type: cli.ParameterTypeBool, DefaultValue: cli.PBool(false), Description: "Do not track the cloned branch."}
)

func init() {
	CSetup.Subcommands["help"] = &cli.Subcommand{
		Description: "Show help for csetup or a specific subcommand",
		Arguments: []cli.Argument{
			cli.Argument{Name: "subcommand", Parameter: HelpSubcommandParameter},
		},
		Exec: func(ctx context.Context, args []string) error {
			return handleHelp(ctx)
		},
	}

	CSetup.Subcommands["init"] = &cli.Subcommand{
		Description: "Initialize a new workspace",
		Arguments: []cli.Argument{
			cli.Argument{Name: "workspace", Parameter: ccommon.WorkspaceParameter},
		},
		RequiredParams: []cli.Parameter{ccommon.WorkspaceParameter},
		Flags:          []cli.Flag{ccommon.ReinitFlag},
		Exec: func(ctx context.Context, args []string) error {
			return handleInit(ctx)
		},
	}
	CSetup.Subcommands["dev-init"] = &cli.Subcommand{
		Description: "Initialize a new workspace for development in the current directory",
		Arguments: []cli.Argument{
			cli.Argument{Name: "workspace", Parameter: ccommon.WorkspaceParameter},
		},
		Flags: []cli.Flag{ccommon.DownloadDepsFlag, ccommon.NoSetupFlag},
		Exec: func(ctx context.Context, args []string) error {
			return handleDevInit(ctx)
		},
	}
	CSetup.Subcommands["git-clone"] = &cli.Subcommand{
		Description: "Clone a git repository into the workspace",
		Arguments: []cli.Argument{
			cli.Argument{Name: "url", Parameter: URLParameter},
			cli.Argument{Name: "source", Parameter: SourceParameter},
		},
		RequiredParams: []cli.Parameter{URLParameter},
		Flags: []cli.Flag{
			ccommon.DownloadDepsFlag,
			ccommon.SubmoduleFlag,
			ccommon.NoSetupFlag,
			cli.Flag{Short: "", Long: "branch", Parameter: BranchParameter},
			cli.Flag{Short: "", Long: "revision", Parameter: RevisionParameter},
			cli.Flag{Short: "", Long: "no-track", Parameter: NoTrackParameter},
		},
		Exec: func(ctx context.Context, args []string) error {
			return handleGitClone(ctx)
		},
	}
	CSetup.Subcommands["declare-git-source"] = &cli.Subcommand{
		Description: "Add git source information to the workspace without downloading",
		Arguments: []cli.Argument{
			cli.Argument{Name: "url", Parameter: URLParameter},
			cli.Argument{Name: "source", Parameter: SourceParameter},
		},
		RequiredParams: []cli.Parameter{URLParameter},
		Exec: func(ctx context.Context, args []string) error {
			return handleDeclareGitSource(ctx)
		},
	}
	CSetup.Subcommands["declare-local-source"] = &cli.Subcommand{
		Description: "Add local source information to the workspace",
		Arguments: []cli.Argument{
			cli.Argument{Name: "source", Parameter: SourceParameter},
			cli.Argument{Name: "sourcepath", Parameter: SourcePathParameter},
		},
		RequiredParams: []cli.Parameter{SourceParameter, SourcePathParameter},
		Flags:          []cli.Flag{SourcePathFlag},
		Exec: func(ctx context.Context, args []string) error {
			return handleDeclareLocalSource(ctx)
		},
	}
	CSetup.Subcommands["download"] = &cli.Subcommand{
		Description: "Download missing sources",
		Arguments: []cli.Argument{
			cli.Argument{Name: "source", Parameter: SourceParameter},
		},
		Flags: []cli.Flag{ccommon.DownloadDepsFlag, ccommon.NoSetupFlag, ccommon.SubmoduleFlag},
		Exec: func(ctx context.Context, args []string) error {
			return handleDownload(ctx)
		},
	}
	CSetup.Subcommands["load-defaults"] = &cli.Subcommand{
		Description: "Load default configuration for a source from its csetup.yml",
		Arguments: []cli.Argument{
			cli.Argument{Name: "source", Parameter: SourceParameter},
		},
		RequiredParams: []cli.Parameter{SourceParameter},
		Exec: func(ctx context.Context, args []string) error {
			return handleLoadDefaults(ctx)
		},
	}
	CSetup.Subcommands["add-dependency"] = &cli.Subcommand{
		Description: "Add a dependency to a target",
		Arguments: []cli.Argument{
			cli.Argument{Name: "target", Parameter: TargetParameter},
			cli.Argument{Name: "dependency", Parameter: DependencyParameter},
		},
		RequiredParams: []cli.Parameter{TargetParameter, DependencyParameter},
		Exec: func(ctx context.Context, args []string) error {
			return handleAddDependency(ctx)
		},
	}
	CSetup.Subcommands["remove-dependency"] = &cli.Subcommand{
		Description: "Remove a dependency from a target",
		Arguments: []cli.Argument{
			cli.Argument{Name: "target", Parameter: TargetParameter},
			cli.Argument{Name: "dependency", Parameter: DependencyParameter},
		},
		RequiredParams: []cli.Parameter{TargetParameter, DependencyParameter},
		Exec: func(ctx context.Context, args []string) error {
			return handleRemoveDependency(ctx)
		},
	}
	CSetup.Subcommands["add-testing-dependency"] = &cli.Subcommand{
		Description: "Add a testing dependency to a target",
		Arguments: []cli.Argument{
			cli.Argument{Name: "target", Parameter: TargetParameter},
			cli.Argument{Name: "dependency", Parameter: DependencyParameter},
		},
		RequiredParams: []cli.Parameter{TargetParameter, DependencyParameter},
		Exec: func(ctx context.Context, args []string) error {
			return handleAddTestingDependency(ctx)
		},
	}
	CSetup.Subcommands["remove-testing-dependency"] = &cli.Subcommand{
		Description: "Remove a testing dependency from a target",
		Arguments: []cli.Argument{
			cli.Argument{Name: "target", Parameter: TargetParameter},
			cli.Argument{Name: "dependency", Parameter: DependencyParameter},
		},
		RequiredParams: []cli.Parameter{TargetParameter, DependencyParameter},
		Exec: func(ctx context.Context, args []string) error {
			return handleRemoveTestingDependency(ctx)
		},
	}
	CSetup.Subcommands["enable-testing"] = &cli.Subcommand{
		Description: "Enable testing for a target",
		Arguments: []cli.Argument{
			cli.Argument{Name: "target", Parameter: TargetParameter},
		},
		RequiredParams: []cli.Parameter{TargetParameter},
		Exec: func(ctx context.Context, args []string) error {
			return handleEnableTesting(ctx)
		},
	}
	CSetup.Subcommands["disable-testing"] = &cli.Subcommand{
		Description: "Disable testing for a target",
		Arguments: []cli.Argument{
			cli.Argument{Name: "target", Parameter: TargetParameter},
		},
		RequiredParams: []cli.Parameter{TargetParameter},
		Exec: func(ctx context.Context, args []string) error {
			return handleDisableTesting(ctx)
		},
	}
	CSetup.Subcommands["remove-source"] = &cli.Subcommand{
		Description: "Remove a source from the workspace",
		Arguments: []cli.Argument{
			ccommon.SourceArg,
		},
		RequiredParams: []cli.Parameter{ccommon.SourceParameter},
		Flags:          []cli.Flag{ccommon.DeleteFlag, ccommon.SourceFlag},
		Exec: func(ctx context.Context, args []string) error {
			return handleRemoveSource(ctx)
		},
	}
	CSetup.Subcommands["remove-target"] = &cli.Subcommand{
		Description: "Remove a target from the workspace",
		Arguments: []cli.Argument{
			ccommon.TargetArg,
		},
		RequiredParams: []cli.Parameter{ccommon.TargetParameter},
		Flags:          []cli.Flag{ccommon.TargetFlag},
		Exec: func(ctx context.Context, args []string) error {
			return handleRemoveTarget(ctx)
		},
	}
	CSetup.Subcommands["remove-project"] = &cli.Subcommand{
		Description: "Remove a source and all its associated targets from the workspace",
		Arguments: []cli.Argument{
			ccommon.SourceArg,
		},
		RequiredParams: []cli.Parameter{ccommon.SourceParameter},
		Flags:          []cli.Flag{ccommon.DeleteFlag, ccommon.SourceFlag},
		Exec: func(ctx context.Context, args []string) error {
			return handleRemoveProject(ctx)
		},
	}
	CSetup.Subcommands["tidy"] = &cli.Subcommand{
		Description: "Delete source folders that are not in the sources list",
		Flags:       []cli.Flag{ccommon.DryRunFlag},
		Exec: func(ctx context.Context, args []string) error {
			return handleTidy(ctx)
		},
	}
	CSetup.Subcommands["new-target"] = &cli.Subcommand{
		Description: "Add a new target to the workspace",
		Arguments: []cli.Argument{
			ccommon.TargetArg,
			ccommon.SourceArg,
		},
		RequiredParams: []cli.Parameter{ccommon.SourceParameter},
		Flags: []cli.Flag{
			ccommon.SourceFlag,
			ccommon.TargetFlag,
			ccommon.OverwriteFlag,
			ccommon.ProjectTypeFlag,
			ccommon.CMakePackageNameFlag,
		},
		Exec: func(ctx context.Context, args []string) error {
			return handleNewTarget(ctx)
		},
	}
	CSetup.Subcommands["set-cxx-version"] = &cli.Subcommand{
		Description: "Set the C++ version for a target or the whole workspace",
		Arguments: []cli.Argument{
			ccommon.CxxVersionArg,
			ccommon.TargetArg,
		},
		RequiredParams: []cli.Parameter{ccommon.CxxVersionParameter},
		Flags:          []cli.Flag{ccommon.TargetFlag},
		Exec: func(ctx context.Context, args []string) error {
			return handleSetCXXVersion(ctx)
		},
	}
	CSetup.Subcommands["enable-staging"] = &cli.Subcommand{
		Description: "Enable staging for a target",
		Arguments: []cli.Argument{
			cli.Argument{Name: "target", Parameter: TargetParameter},
			ccommon.TargetArg,
		},
		RequiredParams: []cli.Parameter{ccommon.TargetParameter},
		Flags:          []cli.Flag{ccommon.TargetFlag},
		Exec: func(ctx context.Context, args []string) error {
			return handleEnableStaging(ctx)
		},
	}
	CSetup.Subcommands["disable-staging"] = &cli.Subcommand{
		Description: "Disable staging for a target",
		Arguments: []cli.Argument{
			cli.Argument{Name: "target", Parameter: TargetParameter},
			ccommon.TargetArg,
		},
		RequiredParams: []cli.Parameter{ccommon.TargetParameter},
		Flags:          []cli.Flag{ccommon.TargetFlag},
		Exec: func(ctx context.Context, args []string) error {
			return handleDisableStaging(ctx)
		},
	}
	CSetup.Subcommands["list-sources"] = &cli.Subcommand{
		Description: "List all sources in the workspace",
		Exec: func(ctx context.Context, args []string) error {
			return handleListSources(ctx)
		},
	}
	CSetup.Subcommands["drop-files"] = &cli.Subcommand{
		Description: "Delete local source files without removing them from configuration",
		Arguments: []cli.Argument{
			cli.Argument{Name: "source", Parameter: SourceParameter},
		},
		Exec: func(ctx context.Context, args []string) error {
			return handleDropFiles(ctx)
		},
	}
	CSetup.Subcommands["get-args"] = &cli.Subcommand{
		Description: "Get build arguments for a target",
		Arguments: []cli.Argument{
			cli.Argument{Name: "target", Parameter: TargetParameter},
			ccommon.TargetArg,
		},
		RequiredParams: []cli.Parameter{ccommon.TargetParameter},
		Flags:          []cli.Flag{ccommon.ConfigFlag, ccommon.ToolchainFlag, ccommon.TargetFlag},
		Exec: func(ctx context.Context, args []string) error {
			return handleGetArgs(ctx)
		},
	}
	CSetup.Subcommands["detect-toolchains"] = &cli.Subcommand{
		Description: "Detect system toolchains",
		Flags:       []cli.Flag{ccommon.DebugFlag},
		Exec: func(ctx context.Context, args []string) error {
			return handleDetectToolchains(ctx)
		},
	}
	CSetup.Subcommands["add-config"] = &cli.Subcommand{
		Description:    "Add a build configuration",
		RequiredParams: []cli.Parameter{ccommon.ConfigParameter},
		Flags:          []cli.Flag{ccommon.ConfigFlag, ccommon.ToolchainFlag},
		Exec: func(ctx context.Context, args []string) error {
			return handleAddConfig(ctx)
		},
	}
	CSetup.Subcommands["remove-config"] = &cli.Subcommand{
		Description:    "Remove a build configuration",
		RequiredParams: []cli.Parameter{ccommon.ConfigParameter},
		Flags:          []cli.Flag{ccommon.ConfigFlag},
		Exec: func(ctx context.Context, args []string) error {
			return handleRemoveConfig(ctx)
		},
	}
	CSetup.Subcommands["pin"] = &cli.Subcommand{
		Description: "Pin a git source (HEAD by default, or a specific revision)",
		Arguments: []cli.Argument{
			cli.Argument{Name: "source", Parameter: SourceParameter},
		},
		Flags: []cli.Flag{
			cli.Flag{Short: "", Long: "revision", Parameter: RevisionParameter},
		},
		Exec: func(ctx context.Context, args []string) error {
			return handlePin(ctx)
		},
	}
	CSetup.Subcommands["track"] = &cli.Subcommand{
		Description: "Track a branch for a git source",
		Arguments: []cli.Argument{
			cli.Argument{Name: "source", Parameter: SourceParameter},
			cli.Argument{Name: "branch", Parameter: BranchParameter},
		},
		RequiredParams: []cli.Parameter{SourceParameter, BranchParameter},
		Exec: func(ctx context.Context, args []string) error {
			return handleTrack(ctx)
		},
	}
	CSetup.Subcommands["untrack"] = &cli.Subcommand{
		Description: "Stop tracking a branch for a git source",
		Arguments: []cli.Argument{
			cli.Argument{Name: "source", Parameter: SourceParameter},
		},
		RequiredParams: []cli.Parameter{SourceParameter},
		Exec: func(ctx context.Context, args []string) error {
			return handleUntrack(ctx)
		},
	}
	CSetup.Subcommands["unpin"] = &cli.Subcommand{
		Description: "Unpin a git source (clear revision)",
		Arguments: []cli.Argument{
			cli.Argument{Name: "source", Parameter: SourceParameter},
		},
		RequiredParams: []cli.Parameter{SourceParameter},
		Exec: func(ctx context.Context, args []string) error {
			return handleUnpin(ctx)
		},
	}
	CSetup.Subcommands["status"] = &cli.Subcommand{
		Description: "Show status for sources (clean/modified, HEAD, expected revision)",
		Arguments: []cli.Argument{
			cli.Argument{Name: "source", Parameter: SourceParameter},
		},
		Exec: func(ctx context.Context, args []string) error {
			return handleStatus(ctx)
		},
	}
	CSetup.Subcommands["versions"] = &cli.Subcommand{
		Description: "List pinned/unpinned versions for sources",
		Arguments: []cli.Argument{
			cli.Argument{Name: "source", Parameter: SourceParameter},
		},
		Exec: func(ctx context.Context, args []string) error {
			return handleVersions(ctx)
		},
	}
	CSetup.Subcommands["update"] = &cli.Subcommand{
		Description: "Update a git source (pull latest or move to revision)",
		Arguments: []cli.Argument{
			cli.Argument{Name: "source", Parameter: SourceParameter},
		},
		RequiredParams: []cli.Parameter{SourceParameter},
		Flags: []cli.Flag{
			cli.Flag{Short: "", Long: "revision", Parameter: RevisionParameter},
		},
		Exec: func(ctx context.Context, args []string) error {
			return handleUpdate(ctx)
		},
	}
}
